#1265 and another typo fix in the same class

alexat opened the pull request:
Optimize available subnets dropdown generation (#1265)
Fix typo in class.Subnets.php

Replaces #1271, did a separate branch to keep different tasks separated

Commit:
Implement different algorithm for subnet dropdown generation

Commit:
Fix potentially problematic typo ($$ instead of $)

Commit:
Pull master

Commit:
Merge https://github.com/phpipam/phpipam

Commit:
Merge https://github.com/phpipam/phpipam

Conflicts:
	functions/classes/class.Subnets.php

Commit:
Merge https://github.com/phpipam/phpipam

Commit:
Merge branch 'master' into optimize-subnet-dropdown

itdependsnetworks commented:
I had put this feature in a few years ago, and sorry for leaving it a mess....

I didn't really go through your code, but I felt like there was some quick wins I never got around to implementing.  Like to not have to look up every subnet. e.g. if /25 was available, you know that every /26, /27, etc... in the block was, and you could peek ahead. Similarly, the reverse is true. 

Anyway, hopefully this is a bit better....

GaryAllan committed:
Bugfix: subnet_dropdown_print_available patch fixes

- Avoid integer overflow on 32bit platforms.
- Ensure $subnets is set to empty list if subnet has no slaves (php error).
- Be consistent with prior behaviour.
  Show all available subnets for the next 8 mask values.
- Remove redundant code.

GaryAllan commented:
Hello @alexat 

I've pushed a few one line bugfixes to your PR branch for the issues found during testing. This fixes generating php errors on subnets with no slaves and issues on 32bit PHP. 

Can you test and report back?

Thanks

Commit:
Merge branch 'optimize-subnet-dropdown' of https://github.com/alexat/phpipam into optimize-subnet-dropdown

Commit:
Continue bugfix: subnet_dropdown_print_available patch fixes

- Completely fix possibility of 32-bit overflows inside optimized code
- Make number of subnets in 'drilldown' range (8 prefix levels) tunable, defaults to 256 (2^8=256, so all for 8 levels)
  Why is that? That's because 256 is already too much and someone may want to display same number for all levels

ghost commented:
Hello @GaryAllan 

Thanks!

I've finally made myself a 32-bit testbed, and supplemented your fixes with some of mine in the last commit, completely fixing possibility of 32-bit signed overflow and making amount of subnets displayed on 'drilldown' and 'lower' levels tunable.

Please test as well.

Commit:
I do not know if phpipam strives to be <5.4 compatible, but changed some [...] to array(...)

itdependsnetworks commented:
I like this `$prefix_lower_limit = 8; # then display only the first 8 availble subnets from the remaining divisions` but I think it should be configurable from the app.

GaryAllan commented:
Hi @alexat 

I don't think we need the $int2pow32 trick as subnet->subnet are stored as strings.  We only have issues if we cast to type integer using gmp_intval instead of gmp_strval.

GaryAllan commented:
Hi @alexat 

Looks like the submodule repos have been accidentally pulled into this PR. I've merged an earlier revision of this PR I've been working on to avoid causing submodule issues. 

Thanks
